modules/common/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/common/transformation/EventUtils.scala

```diff
@@ -93,6 +94,27 @@ object EventUtils {
     tabular.mkString("\t")
   }
 
+  def alterEnrichedEventAny[F[_]: Clock: RegistryLookup: Monad](originalLine: Event, lengths: Map[String, Int]): List[Any] = {
+    def transformDate(s: String): Timestamp =
+      Timestamp.from(Instant.parse(s))
+    def truncate(key: String, value: String): String =
+      lengths.get(key) match {
+        case Some(len) => value.take(len)
+        case None => value
+      }
+
+    originalLine.ordered.flatMap {
+      case ("contexts" | "derived_contexts" | "unstruct_event", _) => None
+      case (key, Some(value)) if key.endsWith("_tstamp") =>
+        val any: Any = value.fold(null, b => b, n => n.toInt.getOrElse(n.toDouble), transformDate, _ => value.noSpaces, _ => value.noSpaces)
+        Some(any)
+      case (key, Some(value)) =>
+        val any: Any = value.fold(null, b => b, n => n.toInt.getOrElse(n.toDouble), s => truncate(key, s), _ => value.noSpaces, _ => value.noSpaces)
```

Comment:
We have some custom decimal fields in SparkSchema, but we use double values in the runtime. I can't see any filled values for such fields in tests, I'm not sure but I think it would result in error. But I also think it could be easily solved with new `FieldValue` structure we going to have in Schema DDL.

Reply:
You mean [this one]( https://github.com/snowplow/snowplow-rdb-loader/blob/feature/databricks-loader/modules/transformer-batch/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/batch/SparkSchema.scala#L33), right ?

From the docs of JsonNumber's `toDouble` method:
```
Convert this number to its best Double approximation.
Anything over Double.MaxValue will be rounded to Double.PositiveInfinity and anything below Double.MinValue is rounded to Double.NegativeInfinity.
```
From [this answer in SO](https://stackoverflow.com/a/36345535): 
```
A double has 15 decimal digits of accuracy if they are all before the decimal point. For numbers with fractional parts you can get many more than 15 digits in the decimal representation, because of the incommensurability of decimal and binary fractions. 
```
Theoretically it is possible to get some number which can get that high but I think it is not very likely in practical sense. Let me know if I am missing something though.

Reply:
I thought on this one more time. I guess it would be better not make any assumption and handle this properly. I will try to extend tests to include this case.

Reply:
I would agree with changing this.  Currently the decimal fields (`tr_total`, `tr_max`, etc) are converted to a double, and then later spark will convert it to a decimal.

The problem is, there are famous example of where double/decimal conversion leads to surprising results (mainly when doing arithmetic, which we're not doing here).  I don't know if this is an example of where a rounding error could occur.  But... let's avoid doubt and treat decimals explicitly.

Reply:
Ignore my comment: I had not realised that the `Event` class already stores those values as doubles, not as something with decimal precision.

I agree, however, that we should at least add tests for extreme double values.  And then possibly fix it if it looks like there is a problem.

Reply:
My main concern is about: `and then later spark will convert it to a decimal`. Is is true? We have doubles in `Event`, we tell spark to expect `DecimalType(18, 2)` explicitly, are we sure it would work (for any double, doesn't have to be extreme)?

Reply:
@pondzix Would be great if you can test this out, or even better write some unit tests.  You are right we should be concerned about mysterious under-the-hood type conversions if we are not in control of them.

Reply:
After [this](https://github.com/snowplow/snowplow-rdb-loader/pull/832/commits/73e1df5137fee9dc8ce834f5ed97e93bf4cd0862) commit tests started to fail with:

`java.lang.RuntimeException: Error while encoding: java.lang.RuntimeException: java.lang.Double is not a valid external type for schema of decimal(18,2)`.